pkg/updatestatus/controlplaneinformer.go

```diff
@@ -64,31 +74,189 @@ func newControlPlaneInformerController(
+			}
+			return err
+		}
+
+		now := c.now()
+		insight := assessClusterOperator(ctx, clusterOperator, targetVersion, c.appsClient, now)
+		msg = makeInsightMsgForClusterOperator(insight, now)
+	default:
+		return fmt.Errorf("invalid queue key %s with unexpected type %s", queueKey, t)
+	}
 	var msgForLog string
 	if klog.V(4).Enabled() {
 		msgForLog = fmt.Sprintf(" | msg=%s", string(msg.insight))
 	}
-	klog.V(2).Infof("CPI :: Syncing ClusterVersion %s%s", clusterVersion.Name, msgForLog)
+	klog.V(2).Infof("CPI :: Syncing %s %s%s", t, name, msgForLog)
 	c.sendInsight(msg)
 
 	return nil
 }
 
+func makeInsightMsgForClusterOperator(coInsight *ClusterOperatorStatusInsight, acquiredAt metav1.Time) informerMsg {
+	uid := fmt.Sprintf("usc-co-%s", coInsight.Name)
+	insight := Insight{
+		UID:        uid,
+		AcquiredAt: acquiredAt,
+		InsightUnion: InsightUnion{
+			Type:                         ClusterOperatorStatusInsightType,
+			ClusterOperatorStatusInsight: coInsight,
+		},
+	}
+	// Should handle errors, but ultimately we will have a proper API and won’t need to serialize ourselves
+	rawInsight, _ := yaml.Marshal(insight)
+	return informerMsg{
+		uid:     uid,
+		insight: rawInsight,
+	}
+}
+
+func assessClusterOperator(ctx context.Context, operator *configv1.ClusterOperator, targetVersion string, appsClient appsv1client.AppsV1Interface, now metav1.Time) *ClusterOperatorStatusInsight {
+	updating := metav1.Condition{
+		Type:               string(ClusterOperatorStatusInsightUpdating),
+		Status:             metav1.ConditionUnknown,
+		Reason:             string(ClusterOperatorUpdatingCannotDetermine),
+		LastTransitionTime: now,
+	}
+
+	imagePullSpec := getImagePullSpec(ctx, operator.Name, appsClient)
+
+	var operatorImage, operatorImageUpgraded, versionUpdated bool
+	for _, version := range operator.Status.Versions {
+		if version.Name == "operator-image" {
+			operatorImage = true
+			if imagePullSpec == version.Version {
+				operatorImageUpgraded = true
+			}
+		}
+		if version.Name == "operator" && version.Version == targetVersion {
+			versionUpdated = true
+		}
+	}
+
+	// "operator-image" might not be implemented by every cluster operator
+	updated := (!operatorImage || operatorImageUpgraded) && versionUpdated
+	if updated {
+		updating.Status = metav1.ConditionFalse
+		updating.Reason = string(ClusterOperatorUpdatingReasonUpdated)
+	}
+
+	var available *configv1.ClusterOperatorStatusCondition
+	var degraded *configv1.ClusterOperatorStatusCondition
+	var progressing *configv1.ClusterOperatorStatusCondition
+
+	for _, condition := range operator.Status.Conditions {
+		condition := condition
+		switch {
+		case condition.Type == configv1.OperatorAvailable:
+			available = &condition
+		case condition.Type == configv1.OperatorDegraded:
+			degraded = &condition
+		case condition.Type == configv1.OperatorProgressing:
+			progressing = &condition
+		}
+	}
+
+	if progressing != nil {
+		if !updated && progressing.Status == configv1.ConditionTrue {
```

Comment:
Looks like we could do `if !updated && progressing != nil` and simplify the conditionals inside?

Reply:
Done.